test/e2e/case18_compliance_api_test.go

```diff
@@ -112,4 +110,758 @@ var _ = Describe("Test policy webhook", Label("compliance-events-api"), Ordered,
+				))
+			})
+		})
+
+		Describe("POST two events on the same cluster and policy", func() {
+			// payload1 defines most things, and should cause the cluster, parent, and policy to be created.
+			payload1 := []byte(`{
+				"cluster": {
+					"name": "cluster4",
+					"cluster_id": "test3-cluster4-fake-uuid-4"
+				},
+				"parent_policy": {
+					"name": "policies.common-parent",
+					"categories": ["cat-3", "cat-4"],
+					"controls": ["ctrl-2"],
+					"standards": ["stand-2"]
+				},
+				"policy": {
+					"apiGroup": "policy.open-cluster-management.io",
+					"kind": "ConfigurationPolicy",
+					"name": "common",
+					"spec": "{\"test\":\"three\",\"severity\":\"low\"}"
+				},
+				"event": {
+					"compliance": "NonCompliant",
+					"message": "configmaps [common] not found in namespace default",
+					"timestamp": "2023-03-03T03:03:03.333Z"
+				}
+			}`)
+
+			// payload2 does not define the parent policy, but it should be inferred because
+			// the policy is the same, even though it only has the spec's hash.
+			payload2 := []byte(`{
+				"cluster": {
+					"name": "cluster4",
+					"cluster_id": "test3-cluster4-fake-uuid-4"
+				},
+				"policy": {
+					"apiGroup": "policy.open-cluster-management.io",
+					"kind": "ConfigurationPolicy",
+					"name": "common",
+					"specHash": "5382228c69c6017d4efbd6e42717930cb2020da0",
+					"severity": "low"
+				},
+				"event": {
+					"compliance": "NonCompliant",
+					"message": "configmaps [common] not found in namespace default",
+					"timestamp": "2023-04-04T04:04:04.444Z"
+				}
+			}`)
+
+			BeforeAll(func(ctx context.Context) {
+				By("POST the events")
+				Eventually(postEvent(ctx, payload1), "5s", "1s").ShouldNot(HaveOccurred())
+				Eventually(postEvent(ctx, payload2), "5s", "1s").ShouldNot(HaveOccurred())
+			})
+
+			It("Should have only created one cluster in the table", func() {
+				rows, err := db.Query("SELECT * FROM clusters WHERE name = $1", "cluster4")
+				Expect(err).ToNot(HaveOccurred())
+
+				count := 0
+				for rows.Next() {
+					var (
+						id        int
+						name      string
+						clusterID string
+					)
+					err := rows.Scan(&id, &name, &clusterID)
+					Expect(err).ToNot(HaveOccurred())
+
+					Expect(id).NotTo(Equal(0))
+					count++
+				}
+
+				Expect(count).To(Equal(1))
+			})
+
+			It("Should have only created one parent policy in a table", func() {
+				rows, err := db.Query("SELECT * FROM parent_policies WHERE name = $1", "policies.common-parent")
+				Expect(err).ToNot(HaveOccurred())
+
+				count := 0
+				for rows.Next() {
+					var (
+						id     int
+						name   string
+						cats   pq.StringArray
+						ctrls  pq.StringArray
+						stands pq.StringArray
+					)
+
+					err := rows.Scan(&id, &name, &cats, &ctrls, &stands)
+					Expect(err).ToNot(HaveOccurred())
+
+					Expect(id).NotTo(Equal(0))
+					count++
+				}
+
+				Expect(count).To(Equal(1))
+			})
+
+			It("Should have only created one policy in a table", func() {
```

Comment:
Actually, this isn't quite right because even though they share similar values, the parent policy and severity are part of the uniqueness constraint, so there should be separate rows.

This stems from my misunderstanding of how nulls are treated in the uniqueness constraints. I'll have a PR up with this fixed.

Reply:
Ok, I think I also had some confusion: I was interpreting omitted fields as "I don't care about these fields, find anything that matches the rest of my request" as opposed to "that field must be empty".